activerecord/lib/active_record/connection_adapters/abstract/query_comment.rb

```diff
@@ -0,0 +1,146 @@
+# frozen_string_literal: true
+
+require "active_support/core_ext/module/attribute_accessors_per_thread"
+
+module ActiveRecord
+  module ConnectionAdapters
+    module QueryComment
+      extend ActiveSupport::Concern
+      included do
+        mattr_accessor :prepend_comment, default: false
+      end
+
+      module ClassMethods
+
+        def inherited(klass)
+          klass.prepend(ExecutionMethods)
+          super(klass)
+        end
+
+        def add_query_comment? # :nodoc:
+          return false unless ActiveRecord::Base.query_comments_enabled
+          return false if QueryCommentContext.components.empty?
+          true
+        end
+
+        def add_query_comment_to_sql(sql) # :nodoc:
+          return sql unless add_query_comment?
+          comment = QueryCommentContext.comment
+          if comment.present? && !sql.include?(comment)
+            sql = if prepend_comment
+              "#{comment} #{sql}"
+            else
+              "#{sql} #{comment}"
+            end
+          end
+          sql
+        end
+      end
+
+      module ExecutionMethods
+        def execute(*args, **kwargs)
+          sql, *rest = args
+          sql = self.class.add_query_comment_to_sql(sql)
+          args = rest.unshift(sql)
```

Comment:
This could be tightened up perhaps:

```ruby
          # sql, *rest = args
          # sql = self.class.add_query_comment_to_sql(sql)
          # args = rest.unshift(sql)
          args[0] = self.class.add_query_comment_to_sql(args[0])
```